src/Pirouette/Term/Transformations.hs

```diff
@@ -295,6 +296,92 @@ constrDestrId = pushCtx "constrDestrId" . rewriteM (runMaybeT . go)
       logTrace $ show tyN
       return $ R.appN (R.appN xCase (map R.Arg xArgs)) (map R.Arg rest)
 
+-- `chooseHeadCase f tyf [st,INPUT,param] INPUT` creates a term which contains the body of `f`
+-- but with a matching on the `INPUT` at the head of it.
+-- For instance, if the input type has two constructors `C1` and `C2`, the output is:
+--
+-- match INPUT#1 with
+--   C1 x -> f st#3 (C1 x#0) param#1
+--   C2 x y -> f st#4 (c2 x#1 y#0) param#2
+
+chooseHeadCase :: (MonadPirouette m)
+               => PrtTerm -> PrtType -> [String] -> String -> m PrtTerm
+chooseHeadCase t ty args fstArg =
+  let argLength = length args in
+  let tyOut = snd (R.tyFunArgs ty) in
+  case elemIndex fstArg args of
+    Nothing -> throwError' $ PEOther "No argument declared as input"
+    Just index ->
+      let tyInput = fst (R.tyFunArgs ty) !! index in
+      case nameOf tyInput of
+        Nothing -> throwError' $ PEOther "The input is not of a pattern-matchable type"
+        Just tyName -> do
+          dest <- blindDest tyOut <$> typeDefOf tyName
+          let transiAbsInput = R.Lam (R.Ann $ fromString "DUMMY_ARG") tyInput $
+                R.appN t (zipWith transitionArgs args [argLength, argLength - 1 .. 1])
+          let body = R.appN dest
+                [ R.Arg $ R.termPure (R.B (fromString fstArg) (toInteger   (argLength - 1 - index)))
+                , R.Arg transiAbsInput
+                ]
+          constrDestrId body
+
+  where
+    nameOf :: PrtType -> Maybe Name
+    nameOf (R.TyApp (R.F (TyFree x)) _) = Just x
+    nameOf _ = Nothing
+
+    -- `blindDest out ty` constructs the function
+    -- \ (i : ty) (f : ty -> out) ->
+    -- match i with
+    --   C1 x0 -> f (C1 x0)
+    --   C2 x0 x1 -> f (C2 x0 x1)
+    blindDest :: PrtType -> PrtTypeDef -> PrtTerm
+    blindDest tyOut (Datatype _ _ dest cons) =
+      R.Lam (R.Ann $ fromString "i") (error "Fictive type of i") $ -- Since the generated term will be applied to two arguments,
+        R.Lam (R.Ann $ fromString "f") (error "Fictive type of f") $ -- those 2 types should never be accessed.
```

Comment:
I understand why you put an error here and I respect it if you rather merge it as is and we fix this later.
That being said, it makes me wonder whether we could use `subst` directly here, making it so that `blindDest` doesn't have to add the lambdas, removing the need for `error`. 

In line `322`, we'd then have: `subst (R.Arg (R.termPure ...) :> R.Arg transiAbsInput :> Id) dest`, and

Reply:
Great idea. I first wrote it as an application when I thought at the algorithm on a piece of paper. But indeed, just writing a substitution avoids to provide dummy annotations of lambdas.